crates/deltalake-core/src/operations/merge/mod.rs

```diff
@@ -1272,25 +1283,38 @@ async fn execute(
     )?;
     new_columns = new_columns.with_column(TARGET_COPY_COLUMN, build_case(copy_when, copy_then)?)?;
 
-    let new_columns = new_columns.into_optimized_plan()?;
+    let new_columns = new_columns.into_unoptimized_plan();
+
+    let distrbute_expr = col(file_column.as_str());
+
+    let merge_barrier = LogicalPlan::Extension(Extension {
+        node: Arc::new(MergeBarrier {
+            input: new_columns,
+            expr: distrbute_expr,
+            file_column,
+        }),
+    });
+
     let operation_count = LogicalPlan::Extension(Extension {
         node: Arc::new(MetricObserver {
             id: OUTPUT_COUNT_ID.into(),
-            input: new_columns,
+            input: merge_barrier,
+            enable_pushdown: false,
         }),
     });
 
     let operation_count = DataFrame::new(state.clone(), operation_count);
     let filtered = operation_count.filter(col(DELETE_COLUMN).is_false())?;
 
     let project = filtered.select(write_projection)?;
-    let optimized = &project.into_optimized_plan()?;
+    let merge_final = &project.into_unoptimized_plan();
```

Comment:
Why is this changed to unoptimized?

Reply:
The conversion to a physical plan will optimize the logical plan so it will save a small amount of work.